rust/rebase/src/flow/nft_ownership.rs

```diff
@@ -0,0 +1,241 @@
+
+    pub async fn process_page(
+        &self,
+        client: &reqwest::Client,
+        u: url::Url,
+        contract_address: &str,
+    ) -> Result<PageResult, FlowError> {
+        let res: AlchemyNftRes = client
+            .get(u)
+            .send()
+            .await
+            .map_err(|e| FlowError::BadLookup(e.to_string()))?
+            .json()
+            .await
+            .map_err(|e| FlowError::BadLookup(e.to_string()))?;
+
+        let mut result: PageResult = PageResult {
+            next_page: res.page_key.clone(),
+            found: false,
+            // res: res.clone(),
+        };
+
+        for nft in res.owned_nfts {
+            if nft.contract.address == contract_address {
+                result.found = true;
+                break;
+            }
+        }
+
+        Ok(result)
+    }
+}
+
+#[async_trait(?Send)]
+impl Flow<Ctnt, Stmt, Prf> for NftOwnership {
+    fn instructions(&self) -> Result<Instructions, FlowError> {
+        Ok(Instructions {
+            statement: "Enter the contract address and network of asset".to_string(),
+            signature: "Sign a statement attesting to ownership of the asset".to_string(),
+            witness: "Send the attestation and the signature to the witness and issue a credential"
+                .to_string(),
+            statement_schema: schema_for!(Stmt),
+            witness_schema: schema_for!(Prf),
+        })
+    }
+
+    async fn statement<I: Issuer>(
+        &self,
+        stmt: &Stmt,
+        _issuer: &I,
+    ) -> Result<FlowResponse, FlowError> {
+        self.sanity_check(&stmt.issued_at)?;
+
+        // TODO: Adjust this when adding additional Alchemy flows.
+        match stmt.subject {
+            Subjects::Pkh(Pkh::Eip155(_)) => {}
+            _ => {
+                return Err(FlowError::Validation(
+                    "Currently only supports Ethereum NFTs".to_string(),
+                ))
+            }
+        }
+
+        Ok(FlowResponse {
+            statement: stmt.generate_statement()?,
+            delimitor: None,
+        })
+    }
+
+    async fn validate_proof<I: Issuer>(&self, proof: &Prf, _issuer: &I) -> Result<Ctnt, FlowError> {
+        self.sanity_check(&proof.statement.issued_at)?;
+
+        let u = Url::parse(&format!(
+            "https://{}-{}.g.alchemy.com/nft/v2/{}/getNFTs?owner={}&withMetadata=false",
+            // TODO: Replace with enum.
+            "eth".to_string(),
+            // TODO: Replace with enum.
+            proof.statement.network,
+            self.api_key,
+            proof.statement.subject.display_id()?
+        ))
+        .map_err(|e| FlowError::BadLookup(e.to_string()))?;
+
+        let client = Client::new();
+        // NOTE: Intentionally re-assigning before read, thus using: #[allow(unused_assignments)]
+        #[allow(unused_assignments)]
+        let mut page_key: Option<String> = None;
+        #[allow(unused_assignments)]
+        let mut res: PageResult = PageResult {
```

Comment:
Don't use a mut, return from the loop

Reply:
?

Reply:
Like `page_key`, `res` is assigned and used in the same loop, so I would do
```rust
     let res = loop {
            let res = self
                .process_page(&client, next_u.clone(), &proof.statement.contract_address)
                .await?;

            if res.found {
                break;
            }

            page_key = res.next_page.clone();
            match page_key {
                None => return res,
                Some(s) => {
                    next_u = u.join(&format!("&pageKey={}", s)).map_err(|_e| {
                        FlowError::BadLookup("Could not follow paginated results".to_string())
                    })?;
                }
            }
        };
```

Reply:
What about `next_u`?

Reply:
In that case it's fine because it's changed in one iteration of the loop and used in the next one. But actually, is this correct? Because it just keeps adding `&pageKey={}` to the URL?

Reply:
If you look close, I keep changing the url that `next_u` is clone from. I never modify the original url, just use it to generate `next_u`.

Reply:
indeed! 😪